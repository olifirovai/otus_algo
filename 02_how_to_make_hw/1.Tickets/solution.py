import datetime as dt
import os

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        answer = int(open(f'{TEST_PATH}\\{test_out[one_case]}', 'r').read())
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', 'r').read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = sum([x ** 2 for x in lucky_tickets(data)])

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def lucky_tickets(n: int, line=[1]):
    new_line = []
    if n > 1:
        line = lucky_tickets(n - 1, line)

    for i in range(9 * n + 1):
        new = 0
        for j in range(max(0, len(line) - 1 - i),
                       min(len(line), 9 * n + 1 - i)):
            new += line[j]
        new_line.append(new)

    # return the last list which is consisted of sums for the half of the
    # whole ticket
    # ex. for n = 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
    return new_line


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
