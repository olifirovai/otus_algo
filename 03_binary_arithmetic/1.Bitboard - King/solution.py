import datetime as dt
import os
from typing import Tuple

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(king_turn(data))
        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def king_turn(position: int) -> Tuple[int, int]:
    king = 1 << position
    no_a = 0xfefefefefefefefe
    no_h = 0x7f7f7f7f7f7f7f7f
    whole_board = 0xffffffffffffffff

    king_a = king & no_a
    king_h = king & no_h
    moves = ((king_a << 7) | (king << 8) | (king_h << 9) |
             (king_a >> 1) | (king_h << 1) |
             (king_a >> 9) | (king >> 8) | (king_h >> 7)) & whole_board

    return count_bits(moves), moves


def count_bits(moves):
    count = 0
    while moves > 0:
        count += 1
        moves &= moves - 1
    return count


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
