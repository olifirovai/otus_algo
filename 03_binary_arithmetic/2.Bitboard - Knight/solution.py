import datetime as dt
import os
from typing import Tuple

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(knight_turn(data))

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(f'data = {data}')
            print(f'answer = {answer}')
            print(f'my result = {result}')
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def knight_turn(position: int) -> Tuple[int, int]:
    knight = 1 << position
    no_a_b = 0xfcfcfcfcfcfcfcfc
    no_b = 0xfdfdfdfdfdfdfdfd
    no_h = 0x7f7f7f7f7f7f7f7f
    no_g = 0xbfbfbfbfbfbfbfbf

    knight_a = knight & no_g & no_h
    knight_b = knight & no_h
    knight_g = knight & no_a
    knight_h = knight & no_a & no_b

    moves = ((knight_b << 17) | (knight_g << 15) |
(knight_a << 10) |                              (knight_h << 6) |
(knight_a >> 10) |                              (knight_h >> 6) |
             (knight_b >> 17) | (knight_g >> 15))

    return count_bits(moves), moves


def count_bits(moves):
    count = 0
    while moves > 0:
        count += 1
        moves &= moves - 1
    return count


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
