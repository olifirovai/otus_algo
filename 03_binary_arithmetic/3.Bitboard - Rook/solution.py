import datetime as dt
import os

from termcolor import colored


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(os.getcwd()):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        answer = int(open(test_out[one_case], "r").read())
        data = int(open(test_in[one_case], "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = rook_turn(data)

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def rook_turn(n: int) -> int:
    pass


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
