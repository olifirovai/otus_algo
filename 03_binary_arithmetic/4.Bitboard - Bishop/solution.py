import datetime as dt
import os
from typing import Tuple

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(bishop_turn(data))

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def bishop_turn(position: int) -> Tuple[int, int]:
    bishop = 1 << position
    no_a = 0xfefefefefefefefe
    no_h = 0x7f7f7f7f7f7f7f7f
    whole_board = 0xffffffffffffffff

    up_left_move = (bishop << 7) & no_h
    up_right_move = (bishop << 9) & no_a
    down_left_move = (bishop >> 9) & no_h
    down_right_move = (bishop >> 7) & no_a

    for _ in range(7):
        up_left_move = (up_left_move | up_left_move << 7) & no_h
        up_right_move = (up_right_move | up_right_move << 9) & no_a
        down_left_move = (down_left_move | down_left_move >> 9) & no_h
        down_right_move = (down_right_move | down_right_move >> 7) & no_a

    moves = (up_left_move | up_right_move | down_left_move | down_right_move) & whole_board

    return count_bits(moves), moves


def count_bits(moves):
    count = 0
    while moves > 0:
        count += 1
        moves &= moves - 1
    return count


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
