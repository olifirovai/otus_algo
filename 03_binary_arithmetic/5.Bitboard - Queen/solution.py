import datetime as dt
import os

from termcolor import colored

TEST_PATH = f'{os.getcwd()}\\tests'


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    for dirpath, dirnames, filenames in os.walk(TEST_PATH):
        for filename in filenames:
            if filename.endswith('in'):
                test_in.append(filename)
            elif filename.endswith('out'):
                test_out.append(filename)

    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    for one_case in range(len(test_in)):
        with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
            answer = [int(line.rstrip('\n')) for line in test_file]
        data = int(
            open(f'{TEST_PATH}\\{test_in[one_case]}', "r").read().strip())

        start_time = dt.datetime.now()

        # here the main function
        result = list(queen_turn(data))

        func_time = (dt.datetime.now() - start_time).total_seconds() * 1000
        output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                      f'finished with {result == answer} result'
        if result == answer:
            print(colored(output_text, 'green'))

        else:
            print(colored(output_text, 'red'))
        print('----------------------------------------------------')


def queen_turn(n: int) -> int:
    pass


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
