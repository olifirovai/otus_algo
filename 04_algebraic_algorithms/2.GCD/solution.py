def euclid_subtracting(data: list) -> int:
    a = data[0]
    b = data[1]
    while a != b:
        while a > b:
            a -= b
        while b > a:
            b -= a
    return a


def euclid_division(data: list) -> int:
    a = data[0]
    b = data[1]
    max_value = max(a, b)
    min_value = min(a, b)
    while max_value != 0 and min_value != 0:
        if max_value % min_value == 0:
            return min_value
        else:
            max_value, min_value = max_value % min_value, max_value


def steinitz_algorithm(data: list) -> int:
    pass
