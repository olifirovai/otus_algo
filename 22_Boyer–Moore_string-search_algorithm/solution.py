class StringSearch:

    def __init__(self, text: str, pattern: str):
        self.text = text
        self.pattern = pattern

    def p_length(self) -> int:
        return len(self.pattern)

    def t_length(self) -> int:
        return len(self.text)

    def find_full_scan(self) -> int:
        t = 0
        while t <= (self.t_length() - self.p_length()):
            p = 0

            while p < self.p_length() and self.text[t + p] == self.pattern[p]:
                p += 1
            if p == self.p_length():
                return t
            t += 1
        return -1

    def create_shift(self) -> list:
        shift = [x for x in range(129)]

        for i in range(len(shift)):
            shift[i] = len(self.pattern)

        for p in range(len(self.pattern) - 1):
            letter_order = ord(self.pattern[p])
            shift[letter_order] = len(self.pattern) - p - 1

        return shift

    def find_jumper(self) -> int:
        shift = self.create_shift()
        for t in range(self.t_length() - self.p_length() + 1):
            p = self.p_length() - 1

            while p >= 0 and self.text[t + p] == self.pattern[p]:
                p -= 1

            if p < 0:
                return t
            t += shift[ord(self.text[t + self.p_length() - 1])]

        return -1

    def boyer_moore_match(self) -> int:
        m = self.p_length()
        n = self.t_length()
        shift = self.create_shift()
        t = 0
        while t <= n - m:
            p = self.p_length() - 1
            while p >= 0 and self.pattern[p] == self.text[t + p]:
                p -= 1
            if p < 0:
                return t
            t += shift[ord(self.text[t + self.p_length() - 1])]

        return -1
