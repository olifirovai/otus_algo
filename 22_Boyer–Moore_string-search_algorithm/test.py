import datetime as dt
import json
import os

from natsort import os_sorted
from termcolor import colored

from solution import StringSearch

TEST_PATH = f'{os.getcwd()}\\tests'
FUNC_DICT = {'find_full_scan': 'Searching with full scan',
             'find_jumper': 'Searching with Jump',
             'boyer_moore_match': 'Boyer-Moore algorithm', }

TEST_TIMES = ['10', '100', '1000', '10000']
TEST_AMOUNT = 8


# find all files with in and out data
def find_test_cases():
    test_in = []
    test_out = []

    filenames = os_sorted(os.listdir(TEST_PATH))
    for filename in filenames:
        if filename.endswith('in'):
            test_in.append(filename)
        elif filename.endswith('out'):
            test_out.append(filename)
    return test_in, test_out


def make_tests(test_in: list, test_out: list) -> None:
    if 'result_dict.txt' not in os.listdir(os.getcwd()):
        create_dict_txt(test_in, FUNC_DICT)

    for func in FUNC_DICT.keys():
        start_text = f'Tests with function {func}'
        print(colored(start_text, 'magenta', attrs=['bold']))
        print('----------------------------------------------------')

        for one_case in range(len(test_in)):
            with open(f'{TEST_PATH}\\{test_in[one_case]}', "r") as test_file:
                value_list = test_file.readlines()
                text_value = value_list[0][:-1]
                pattern_value = value_list[1]
            with open(f'{TEST_PATH}\\{test_out[one_case]}', "r") as test_file:
                answer = int(test_file.read())

            start_time = dt.datetime.now()

            # here the main functions
            for once in TEST_TIMES:
                for _ in range(int(once)):
                    search_node = StringSearch(text_value, pattern_value)
                    result = eval(f'search_node.{func}()')
                func_time = (
                                    dt.datetime.now() - start_time).total_seconds() * 1000

                load_data_txt(one_case, FUNC_DICT, func, func_time, once)

                output_text = f'{test_in[one_case][:-3]} run with {func_time} ms, ' \
                              f'finished with {result == answer} result'
                if result == answer:
                    print(colored(output_text, 'green'))

                else:
                    print(f'my result {result}')
                    print(f'answer    {answer}')
                    print(colored(output_text, 'red'))
            print('----------------------------------------------------')
    print_table_readme('result_dict.txt')


def create_dict_txt(test_in, func_dict):
    with open('result_dict.txt', 'w') as s:
        result_dict = {}
        for func in func_dict.values():
            result_dict[func] = {}
            for amount in TEST_TIMES:
                result_dict[func][amount] = {}
                for test in range(len(test_in) // len(TEST_TIMES)):
                    result_dict[func][amount][str(test)] = 'Too much'
        s.write(json.dumps(result_dict))


def load_data_txt(one_case, func_dict, func, func_time, test_amount):
    with open('result_dict.txt', 'r') as f:
        data = f.read()
        result_dict = json.loads(data)

    with open('result_dict.txt', 'w') as s:
        test_num = f'{one_case}'
        func_disc = func_dict[func]
        result_dict[func_disc][test_amount][test_num] = func_time
        s.write(json.dumps(result_dict))


def print_table_readme(file):
    with open(file, 'r') as f:
        data = f.read()
        result_dict = json.loads(data)

    list_sort = list(result_dict.keys())
    header = '|Times|Test â„–|'
    header_line = '|---|---|'
    for one_sort in list_sort:
        header += f'{one_sort}|'
        header_line += '---|'
    print(header)
    print(header_line)
    for test in range(TEST_AMOUNT):
        for once in TEST_TIMES:
            line = f'|{once}|{test}'
            for sort in list_sort:
                time = result_dict[sort][once][str(test)]
                line += f'|{round(time, 2)} ms'
            print(f'{line}|')


def main():
    test_in, test_out = find_test_cases()
    make_tests(test_in, test_out)


if __name__ == '__main__':
    main()
