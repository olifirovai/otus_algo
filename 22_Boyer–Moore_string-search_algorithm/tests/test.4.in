version https://git-lfs.github.com/spec/v1
oid sha256:de5821f385a113bb26c1dae3bc61979e41210c4e758f4c29a3b5b3f31a174f15
size 1119
