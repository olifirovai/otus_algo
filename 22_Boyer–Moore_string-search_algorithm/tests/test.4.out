version https://git-lfs.github.com/spec/v1
oid sha256:3097fc802d49355a0a256d4b07ab9f7257fcf35077cdb8133d59f607fedb229e
size 3
